tests/test_runlib.py

```diff
@@ -181,8 +179,98 @@ def test_record_dot_check_files_hash_dict_schema(self):
       securesystemslib.formats.HASHDICT_SCHEMA.check_match(val)
 
     self.assertListEqual(sorted(list(artifacts_dict.keys())),
-      sorted(["foo", "bar", "subdir/foosub1", "subdir/foosub2",
-          "subdir/subsubdir/foosubsub"]))
+      sorted(self.full_file_path_list))
+
+
+  def test_record_symlinked_files(self):
+    """Symlinked files are always recorded. """
+    # Symlinked **files** are always recorded ...
+    link_pairs = [
+      ("foo", "foo_link"),
+      ("subdir/foosub1", "subdir/foosub2_link"),
+      ("subdir/subsubdir/foosubsub", "subdir/subsubdir/foosubsub_link")
+    ]
+
+    # Create links
+    for pair in link_pairs:
+      # We only use the basename of the file (source) as it is on the same
+      # level as the link (target)
+      os.symlink(os.path.basename(pair[0]), pair[1])
+
+    # Record files and linked files
+    # follow_links does not make a difference as it only concerns linked dirs
+    for follow_links in [True, False]:
+      artifacts_dict = record_artifacts_as_dict(["."], follow_links)
+
+      # Test that everything was recorded ...
+      self.assertListEqual(sorted(list(artifacts_dict.keys())),
+          sorted(self.full_file_path_list + [pair[1] for pair in link_pairs]))
+
+      # ... and the hashes of each link/file pair match
+      for pair in link_pairs:
+        self.assertDictEqual(artifacts_dict[pair[0]], artifacts_dict[pair[1]])
+
+    for pair in link_pairs:
+      os.unlink(pair[1])
+
+
+  def test_record_without_dead_symlinks(self):
+    """Dead symlinks are never recorded. """
+
+    # Dead symlinks are never recorded ...
+    links = [
+      "foo_link",
+      "subdir/foosub2_link",
+      "subdir/subsubdir/foosubsub_link"
+    ]
+
+    # Create dead links
+    for link in links:
+      os.symlink("does/not/exist", link)
+
+    # Record files without dead links
+    # follow_links does not make a difference as it only concerns linked dirs
+    for follow_links in [True, False]:
+      artifacts_dict = record_artifacts_as_dict(["."], follow_links)
+
+      # Test only the files were recorded ...
+      self.assertListEqual(sorted(list(artifacts_dict.keys())),
+          sorted(self.full_file_path_list))
+
+    for link in links:
+      os.unlink(link)
+
+
+  def test_record_follow_symlinked_directories(self):
+    """Record files in symlinked dirs if follow_links is True. """
```

Comment:
I don't recall whether the docstring for `follow_links` explains it, but which file path is recorded when a symlinked directory is visited?  The original path, or the resolved one?